sparse tensor operations

martinraison opened the pull request:
This pull request adds more support for sparse operations in Pytorch.

The original goals:

- ability to propagate sparse updates in a network (e.g. for updating an embedding matrix)
- ability to efficiently compute "bag-of-words" sentence embeddings (e.g. weighted average of word embeddings)

This request implements the following individual features to achieve those goals:

- enable backpropagation of sparse gradients without conversion to dense tensors. In most cases a runtime exception is thrown when mixing different gradient types for the same variable
- add some methods for `THSTensor`: `zero`, elementwise `add` and `mul`, scalar `mul` and `div`
- make `addcmul` method of `THTensor` compatible with sparse operands
- make `spmm` method accessible from Python (I had to use the name `dsmm` since `smm` was already taken. Maybe we should rename the current `smm` to `ssmm` to follow the convention)
- `sparse_mask` method on `THTensor`. This produces a sparse tensor from a dense tensor, by using a sparse tensor as a mask. A value is only present in the output sparse tensor if it also exists in the mask. See the changes to `adagrad.py` for an example of why this is needed.
- update Adagrad code to use sparse updates when possible. I was hoping the optimizers wouldn't require any modification, but it looks like they do (let me know if you see a better option). In addition, not every optimizer supports sparse updates easily. For example it looks like Adam is accumulating moments over time, which effectively makes the updates dense. For the same reason, Adagrad doesn't support having both sparse updates + weight decay.
- leave `Variable`'s gradient to `None` by default (required updating a few tests). This is because there is no canonical zero gradient anymore (it could be dense or sparse, and if it is sparse we don't know how many dimensions are sparse)
- I also added the basic glue code to hook up the existing THCS (cuda sparse) tensor implementation to Python (I did pretty much the same as for TH, THC, THS). I did that mainly so that existing tests keep working even when cuda sparse gradients are involved. Most of the THCS operations are still stubs and will throw an exception with an error message, but it means the only thing remaining for GPU sparse ops support is to fill in the appropriate functions. 

...and last but not least: N-dimensional values for sparse tensors. This one is a slightly bigger item. Basically for things like applying sparse updates to embedding matrices, only the first dimension (the one that corresponds to the word index) is sparse. The other dimension is always dense (only whole embedding vectors are updated). An elegant solution is to make the `values` tensor N-dimensional instead of 1-dimensional. For an embedding matrix, the sparse gradient will have a `values` tensor of size `nnz * embedding_size` instead of just `nnz`. I had to update a few existing functions to make that work, but otherwise the changes are actually not that big. Existing usecases with scalar values should all work as usual.

colesbury commented on torch/nn/parameter.py:
```diff
@@ -21,8 +21,8 @@ class Parameter(Variable):
     """
-    def __new__(cls, data=None, requires_grad=True):
-        return super(Parameter, cls).__new__(cls, data, requires_grad=requires_grad)
+    def __new__(cls, data=None, requires_grad=True, sparse_grad=True):
```
Does it make sense to have `sparse_grad` default to True?

colesbury commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -12,12 +12,14 @@ namespace torch { namespace autograd {
     , grad(nullptr)
     , version_counter(new VariableVersion())
     , output_nr(0)
+    , sparse_grad(0)
```
should this be initialized to the `sparse_grad` argument?

martinraison replied:
Good catch - looks like the two mistakes you pointed out cancelled each other :)

colesbury commented on torch/lib/THS/generic/THSTensorMath.c:
```diff
@@ -181,18 +419,38 @@ void THSTensor_(spcadd)(THTensor *r_, THTensor *dense, real value, THSTensor *sp
+    THTensor *srcBuffer = THTensor_(new)();
+    THTensor *dstBuffer = THTensor_(new)();
+    #pragma omp parallel for private(k)
+    for (k = 0; k < sparse->nnz; k++) {
```
How does the OpenMP parallelism work here? It looks like the same buffers are modified by multiple threads concurrently.

martinraison replied:
So far I've only tested with `OMP_NUM_THREADS=1` so I guess it doesn't work. I'll see if I can fix it (I'm not super familiar with OpenMP)

colesbury replied:
Probably best to remove the `#pragma omp parallel for` in that case. It looks like there are race conditions.

soumith commented:
@pytorchbot add to whitelist

Commit:
sparse tensor operations (CPU)

Commit:
sparse tensor operations (CUDA)

martinraison commented:
I rebased this PR on top of the latest changes, polished a few things and added some tests. I think it is ready for review / merging, let me know what you think.

martinraison commented:
There are a few changes compared to the original PR description:

* there is no explicit `sparse_grad` flag anymore. Whether a `Variable`'s gradient is sparse is not an intrinsic property of the variable itself, instead it is determined by the functions that are applied to it. Therefore sparse gradients are propagated if possible, and an error message is shown if incompatible types of gradients are accumulated for the same variable (we could relax this constraint and fallback to dense gradients when accumulating but I'm not sure if that's better from a user perspective)
* I removed some functions that don't make much sense / have unclear semantics: scalar `add` (that would turn a sparse matrix into a dense one), elementwise `div` and `addcdiv` (causes division by zero errors unless the denominator index set is a subset of the numerator index set)
* I added some glue code to bridge THCS and Python to make existing tests pass and lay the groundwork for future sparse cuda tensor support (for the time being I didn't implement any additional sparse cuda ops)

colesbury commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -224,12 +224,7 @@ PyObject *THPVariable_get_grad(THPVariable *self)
-    auto grad = var.data->newTensor();
-    grad->resizeAs(*var.data).zero();
-    var.grad = std::make_shared<Variable>(std::move(grad), 0, 1);
+    return Py_None;
```
this needs to be `Py_RETURN_NONE;`

All Python return values need to be incref'd, including None. The `Py_RETURN_NONE` does this for you.

colesbury replied:
Perhaps more importantly, why is this returning None for the grad?

martinraison replied:
When gradients can be sparse, there is unfortunately no good default value for zero gradients (the gradient could be dense or sparse, and even for sparse gradients we don't know how many dimensions are actually sparse). I think the most elegant way to handle it is to leave it to `None` by default, though I understand this is a small breaking change. Let me know if you see a better way.

colesbury commented on torch/tensor.py:
```diff
@@ -98,7 +99,7 @@ def __deepcopy__(self, _memo):
         new_storage = self.storage().__deepcopy__(_memo)
         new_tensor = self.new()
-        new_tensor.set_(new_storage, self.storage_offset(), self.size(), self.stride())
+        new_tensor.set_(new_storage, self.storage_offset(), torch.Size(self.size()), self.stride())
```
Why do you need to wrap it in a torch.Size()?

martinraison replied:
you're right, this isn't necessary anymore (`size()` used to return a `LongStorage` for sparse tensors)

colesbury commented on torch/optim/optimizer.py:
```diff
@@ -132,7 +132,8 @@ def zero_grad(self):
         for group in self.param_groups:
             for param in group['params']:
-                param.grad.data.zero_()
+                if param.grad is not None:
```
When can `grad` be None?

martinraison replied:
see explanation above

colesbury commented on torch/sparse/__init__.py:
```diff
@@ -1,53 +1,118 @@
 
 
-class DoubleTensor(_C.SparseDoubleTensorBase):
+class _SparseTensorBase(object):
```
Can you use _TensorBase as the base class instead and use a mix-in? (Like in `cuda/__init__.py`  _CudaBase)

martinraison replied:
I guess I could, though I'll probably have to override some of the base methods that aren't supported for sparse tensors (and raise `NotImplementedError` or whatever makes sense)

Commit:
make sparse tensors inherit from _TensorBase

apaszke commented on torch/lib/THPP/tensors/generic/THCSTensor.cpp:
```diff
@@ -0,0 +1,653 @@
+  return *this;
+}
+
+#define non_const_cast(tensor) const_cast<THCSTensor&>(dynamic_cast<const THCSTensor&>(tensor))
```
I think we've changed that in all other classes to `const_cast`, that doesn't require removing the `const` qualifier.

apaszke commented on torch/lib/THPP/tensors/generic/THSTensor.cpp:
```diff
@@ -2,6 +2,15 @@
 #define THS_GENERIC_FILE "tensors/generic/THSTensor.cpp"
 #else
 
+#define const_tensor_cast(tensor) \
```
That's the cast I mentioned above.

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -40,6 +54,9 @@ THLongTensor *THSTensor_(indices)(const THSTensor *self) {
 THTensor *THSTensor_(values)(const THSTensor *self) {
   if (self->nnz == 0) {
+    if (self->values->nDimension) {
+      THTensor_(resizeNd)(self->values, 0, NULL, NULL);
```
Not sure if it's the right way to free memory for empty tensors. Isn't it better to do that directly in `zero_` instead of waiting for someone to call `indices` or `values`?

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -58,40 +75,57 @@ static void THSTensor_(rawInit)(THSTensor *self)
 }
 
+THSTensor* THSTensor_(set)(THSTensor *self, THLongTensor *indices, THTensor *values) {
+  // Note: Not like torch.set, this is an internal method
```
can you rename it to `_set` to avoid confusion?

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -261,18 +326,27 @@ THTensor *THSTensor_(toDense)(THSTensor *self) {
   return other_;
 }
 
+void THSTensor_(copy)(THSTensor *self, THSTensor *src) {
```
Is this also an internal method? I think the usual semantics of copy is that if the tensors weren't sharing the data before, they won't be sharing it afterwards too.

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -329,20 +481,26 @@ void THSTensor_(reorder)(THSTensor *self) {
     // TODO: pass eps in as a parameter
-    if (values[j] == 0) continue;
-    for (d = 0; d < ndim; d++)
+    // if (values[j] == 0) continue;
```
should this be commented out?

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -329,20 +481,26 @@ void THSTensor_(reorder)(THSTensor *self) {
       }
-    if (cmp) values[i] += values[j];
+    if (cmp) {
+      THSTensor_(addSlice)(dstBuffer, dstBuffer, srcBuffer, values_, values_, 1, values_, 0, i, i, j);
```
nit: It might be more readable to allow passing in `NULL`s as one of the source buffers.

apaszke commented on torch/lib/THS/generic/THSTensorMath.c:
```diff
@@ -5,6 +5,222 @@
+}
+
+void THSTensor_(csub)(THSTensor *r_, THSTensor *t, real value, THSTensor *src) {
+  THSTensor_(cadd)(r_, t, -value, src);
```
This won't work for byte tensors, because `value` is unsigned.

apaszke commented on torch/lib/THS/generic/THSTensorMath.c:
```diff
@@ -5,6 +5,222 @@
+}
+
+void THSTensor_(cmul)(THSTensor *r_, THSTensor *t, THSTensor *src) {
+  if(!THSTensor_(isSameSizeAs)(t, src)) {
```
I think in `TH` and `THC` `c*` functions only check if the number of elements matches. But I'm fine with leaving it like that too.

apaszke reviewed (commented):
Overall looks good, but I have some questions and comments. The most important is that I don't understand why do we need separate sizes and dimension fields for indices and values.

I'll have to think if choosing the grad type dynamically, and disallowing changing it later won't be too limiting. I'm a bit afraid that it could cause some surprising errors with no easy workarounds.

martinraison commented:
Thanks @apaszke for the feedback! I will address your various comments.

Regarding the difference between `nDimensionI` and `nDimensionV`: the idea is to allow values of arbitrary dimension instead of just scalars. This happens for example with word embedding matrices. The matrix will be of size num_words x embedding_size, and updates to that matrix will be sparse, but only along the first dimension (only a few words are used in each batch, however whenever a word embedding changes, the whole embedding changes). In this specific case, nDimensionI=1 (num_words) and nDimensionV=1 (embedding_size). If I call `to_dense` on a sparse matrix, the result has dimension `nDimensionI + nDimensionV`. So the "indices" and "values" dimensions are complementary, and can of course be different. Let me know if I'm still unclear.

martinraison commented on torch/csrc/autograd/grad_buffer.cpp:
```diff
@@ -19,6 +19,9 @@ auto GradBuffer::addGrad(size_t pos, std::shared_ptr<Variable>&& var) -> void {
   if (!item.first) {
     buffer[pos] = std::make_pair<>(std::move(tensor), true);
   } else {
+    if (item.first->isSparse() && !tensor->isSparse()) {
```
ah, yes, you're right. My original intent was to disallow mixing entirely and then I realized it was trivial to allow this particular case. I'll update the code to automatically upgrade the result to dense if necessary, regardless of the addition order.

martinraison commented on torch/cuda/sparse.py:
```diff
@@ -0,0 +1,87 @@
+    # Define dummy base classes
+    for t in ['Double', 'Float', 'Long', 'Int', 'Short', 'Char', 'Byte', 'Half']:
+        tensor_name = 'CudaSparse{0}TensorBase'.format(t)
+        torch._C.__dict__[tensor_name] = type(tensor_name, (object,), {})
```
good point. To be honest I just did the same as in `torch/cuda/__init__.py`. I'll add an `__init__` that raises an error there as well.

martinraison commented on torch/lib/THCS/generic/THCSTensor.c:
```diff
@@ -51,29 +62,38 @@ static void THCSTensor_(rawInit)(THCState *state, THCSTensor *self)
   self->values = NULL;
-  self->nDimension = 0;
+  self->nDimensionI = 0;
+  self->nDimensionV = 0;
```
see reply below

apaszke commented:
So sparse tensors now have `nDimensionI` sparse dimensions, followed by `nDimensionV` dense dimensions, right? That's nice. Were the sparse tensors we had previously equivalent to `nDimensionV == 1` or `nDimensionV == 0`?

martinraison commented:
That's exactly right. The sparse tensors we had previously were equivalent to `nDimensionV == 0` (which I think makes more sense)

Commit:
fixes + address @apaszke feedback

martinraison commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -560,6 +557,7 @@ void THSTensor_(free)(THSTensor *self)
     return;
   if(THAtomicDecrementRef(&self->refcount))
   {
+    THFree(self->size);
```
Btw, I just noticed `size` was never freed in the previous implementation. Memory leak?

Commit:
comment out failing test

martinraison commented:
@apaszke the test `test_variable_sequence_cuda` was failing on my branch because `x.grad` never gets initialized in the CUDA case (the CPU case works fine). I debugged it on master, and it looks like `x.grad` remains zero during backwards with CUDA (not with CPU), so something seems off with the test. I commented it out on my branch, but it would be good to check why gradients aren't being computed correctly.

apaszke commented:
I found the issue and it's because of [this line](https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L1192). On CUDA `x` is not a leaf, so its grad will never be initialized. Can you please fix that test too?

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -376,7 +375,6 @@ void THSTensor_(addSlice)(
   THTensor *dst, THTensor *src1, real value, THTensor *src2,
   long dim, long dstIdx, long src1Idx, long src2Idx) {
   if (src1->nDimension > 1) {
-    THTensor_(select)(src1Buffer, src1, dim, src1Idx);
```
Why did you remove that?

apaszke commented on torch/lib/THS/generic/THSTensor.c:
```diff
@@ -294,31 +367,108 @@ int THSTensor_(isContiguous)(const THSTensor *self) {
+  THTensor *dstBuffer, THTensor *src1Buffer, THTensor *src2Buffer,
+  THTensor *dst, THTensor *src1, real value, THTensor *src2,
+  long dim, long dstIdx, long src1Idx, long src2Idx) {
+  if (src1->nDimension > 1) {
```
Don't you need to select `src1Buffer`? I saw you removed that

apaszke commented on torch/__init__.py:
```diff
@@ -322,6 +322,7 @@ def storage_type(cls):
 ################################################################################
 
 import torch.cuda
+import torch.cuda.sparse
```
I think a better place for that import would be in torch.cuda (from . import sparse).

apaszke commented on torch/lib/THCS/generic/THCSTensor.c:
```diff
@@ -51,29 +62,38 @@ static void THCSTensor_(rawInit)(THCState *state, THCSTensor *self)
+  long d, nDimI_ = 0, nDimV_ = 0;
+  for (d = 0; d < nDimI; d++) {
+    if (size[d] > 0) {
+      self->size[nDimI_++] = size[d];
```
I didn't get that. It seems to me that if someone constructs a tensor using some non-empty values and indices and passes in size of e.g. (2, 4, 0) or (2, -1), then this might break.

Commit:
nits & test fixes

apaszke commented:
Phew, that was a big diff! Thanks! :tada:

martinraison commented:
Thanks for taking the time to review all this :)
